sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/AbstractRulesDefinition.java

```diff
@@ -86,26 +91,44 @@ public void define(Context context) {
     RulesDefinitionXmlLoader loader = new RulesDefinitionXmlLoader();
     loader.load(repository, new InputStreamReader(getClass().getResourceAsStream(rulesXmlFilePath), StandardCharsets.UTF_8));
 
-    for (NewRule rule : repository.rules()) {
-      updateMetadata(rule);
-    }
+    setupHotspotRules(repository.rules());
 
     repository.done();
   }
 
-  private void updateMetadata(NewRule rule) {
+  private void setupHotspotRules(Collection<NewRule> rules) {
+    Map<NewRule, RuleMetadata> allRuleMetadata = rules.stream()
+      .collect(Collectors.toMap(rule -> rule, rule -> readRuleMetadata(rule.key())));
+
+    Set<NewRule> hotspotRules = getHotspotRules(allRuleMetadata);
+
+    // Either set security standards fields, or remove the rules altogether,
+    // depending on whether the SonarQube instance supports hotspots or not.
     if (supportsSecurityHotspots) {
-      RuleMetadata ruleMetadata = readRuleMetadata(rule.key());
-      for (String s : ruleMetadata.securityStandards.OWASP) {
-        rule.addOwaspTop10(RulesDefinition.OwaspTop10.valueOf(s));
-      }
-      rule.addCwe(ruleMetadata.securityStandards.CWE);
-      if (ruleMetadata.isSecurityHotspot()) {
-        setSecurityHotspotType(rule);
+      for (Map.Entry<NewRule, RuleMetadata> entry : allRuleMetadata.entrySet()) {
```

Comment:
nitpick: you can use inline forEach for this as well, it would improve readability

    allRuleMetadata.forEach(AbstractRulesDefinition::updateSecurityStandards);